docs/Dialects/Interop/RationaleInterop.md

```diff
@@ -0,0 +1,590 @@
+# Interoperability Dialect Rationale
+
+This document describes the various design points of the Interop dialect, a
+dialect that is used to represent partially lowered interoperability layers and
+that provides common interfaces and utilities for automated interop generation.
+This follows in the spirit of other
+[MLIR Rationale docs](https://mlir.llvm.org/docs/Rationale/).
+
+- [Interoperability Dialect Rationale](#interoperability-dialect-rationale)
+  - [Introduction](#introduction)
+  - [Procedural Interop](#procedural-interop)
+    - [Representing Partial Lowerings](#representing-partial-lowerings)
+    - [Interop Mechanisms](#interop-mechanisms)
+    - [Instance-side Lowering](#instance-side-lowering)
+    - [Container-side Lowering](#container-side-lowering)
+    - [Bridging between Interop Mechanisms](#bridging-between-interop-mechanism)
+    - [How to use this dialect](#how-to-use-this-dialect)
+    - [Design considerations](#design-considerations)
+
+## Introduction
+
+The increasing number of CIRCT-based backends as well as the usage of many
+external tools raises questions about interoperability and composition of
+those tools. Are we supposed to write interop-layers between them by hand or
+can this also be automatically generated? Can we generate interfaces to the
+outside in multiple standardized formats/languages automatically?
+
+Let's pick simulation as a concrete example. We have Verilator, CIRCT-based
+simulators like the event-driven `llhd-sim`, and at some point maybe another
+non-event-driven simulator to compete with Verilator in terms of performance.
+They all provide their own interface for writing test-benches in a way that
+suits them best. Verilator has its custom C++ interface as well as SystemC
+support, `llhd-sim` expects the test-bench to be written in LLHD (e.g., compiled
+from SystemVerilog), and nobody knows what the interface of the third would look
+like (maybe a custom C-header?). This means, if we want to simulate a design
+with all three, we have to write the test-bench three times or at least some
+layer between the TB and the simulator interfaces. This dialect aims to solve
+this issue by auto-generating an interface from a selection of standarized
+formats (e.g., a SystemC or SystemVerilog module) such that the test-bench has
+to be written only once and switching to another simulator just means changing
+one CLI argument.
+What if we have to use some blackbox Verilog code in the design? Is it compiled
+by a (not yet existent) SystemVerilog front-end and simulated by the CIRCT-based
+simulators or could it also be useful to simulate the blackbox using Verilator
+and everything surrounding it using the CIRCT-based simulator? This dialect can
+provide the latter by implementing an interop operation and one lowering per
+simulator rather than custom pair-wise interop implementations.
```

Comment:
Nice. The PR is good to explicit mention that this is to avoid the `N x M` problem.